crates/admin/src/cluster_controller/scheduler.rs

```diff
@@ -153,303 +117,222 @@ impl<T: TransportConnect> Scheduler<T> {
-                        target_state.leader = None;
-                        modified = true;
+                    if target_state.node_set.len() > replication_factor {
+                        for node_id in target_state
+                            .node_set
+                            .iter()
+                            .filter(|node_id| Some(**node_id) != preferred_leader)
+                            .cloned()
+                            .choose_multiple(
+                                &mut rng,
+                                target_state.node_set.len() - replication_factor,
+                            )
+                        {
+                            target_state.node_set.shift_remove(&node_id);
+                        }
                     }
                 }
+            }
+        }
 
-                modified
-            })
+        // check if the leader is still part of the node set; if not, then clear leader field
+        if let Some(leader) = target_state.leader.as_ref() {
+            if !target_state.node_set.contains(leader) {
+                target_state.leader = None;
+            }
         }
     }
 
-    fn ensure_leadership(
+    fn ensure_leadership<H: PartitionProcessorPlacementHints>(
         &self,
-        scheduling_plan_builder: &mut SchedulingPlanBuilder,
-        placement_hints: impl PartitionProcessorPlacementHints,
+        partition_id: &PartitionId,
+        target_state: &mut TargetPartitionReplicationState,
+        placement_hints: &H,
     ) {
-        let partition_ids: Vec<_> = scheduling_plan_builder.partition_ids().cloned().collect();
-        for partition_id in partition_ids {
-            scheduling_plan_builder.modify_partition(&partition_id, |target_state| {
-                let preferred_leader = placement_hints.preferred_leader(&partition_id);
-                if target_state.leader.is_none() {
-                    target_state.leader =
-                        self.select_leader_from(&target_state.node_set, preferred_leader);
-                    // check whether we modified the leader
-                    return target_state.leader.is_some();
-                } else if preferred_leader.is_some_and(|preferred_leader| {
-                    Some(preferred_leader) != target_state.leader
-                        && target_state.node_set.contains(&preferred_leader)
-                }) {
-                    target_state.leader = preferred_leader;
-                    return true;
-                }
-
-                false
-            })
+        let preferred_leader = placement_hints.preferred_leader(partition_id);
+
+        if target_state.leader.is_none() {
+            target_state.leader = self.select_leader_from(target_state, preferred_leader);
+            // check whether we modified the leader
+        } else if preferred_leader.is_some_and(|preferred_leader| {
+            Some(preferred_leader) != target_state.leader
```

Comment:
I think that this check here `Some(preferred_leader) != target_state.leader` should no longer be needed because we don't return a value from this branch anymore.

Reply:
I don't get what you mean. The target_state.leader can still be set to `None` by the `ensure_replication` function. This can happen for example if the leader node died the ensure_replication will make sure the nodeset has an updated set but will not choose a leader. The leader will then be setup here

Reply:
With this check we only assign the preferred leader if it is different from `target_state.leader`. W/o it we would assign it uncoditionally of this aspect (simplifying the if condition a bit).